<?xml version="1.0" ?>
<robot name="vehicle_4ws" xmlns:xacro="http://www.ros.org/wiki/xacro">

<xacro:property name="PI" value="3.1415926535897932384626433832795028841971" />
<xacro:property name="gravitational_acceleration" value="9.80665" />
<xacro:property name="vehicle1_Yaw" value="-PI/4" />

<!-- 荷台と連結点のmass -->
<xacro:property name="fusion_vehicle_body_mass" value="2113.272311120373" /> 
<xacro:property name="vehicle_body_mass" value="100" /> 
<xacro:property name="vehicle_inertia_coef" value="${vehicle_body_mass/fusion_vehicle_body_mass}" /> 

<xacro:property name="wheel_mass" value="4.639715839283939" /> 

<!-- frontは受動車輪なので低めに設定 -->
<xacro:property name="front_wheel_joint_damping" value="1.0"/>
<xacro:property name="front_wheel_joint_mu" value="0.1" />
<xacro:property name="front_wheel_joint_friction" value="${front_wheel_joint_mu* gravitational_acceleration * wheel_mass  }" /> 
<xacro:property name="front_steering_damping" value="3.0" />
<xacro:property name="front_steering_mu" value="0.001" />
<xacro:property name="front_steering_friction" value="${front_steering_mu * gravitational_acceleration * 0.25*vehicle_body_mass}" />

<!-- rearは駆動輪なので、入力以外で動かないように設定 -->
<xacro:property name="rear_wheel_joint_damping" value="1.0"/>
<xacro:property name="rear_wheel_joint_mu" value="0.1" />
<xacro:property name="rear_wheel_joint_friction" value="${rear_wheel_joint_mu* gravitational_acceleration * wheel_mass }" /> 
<xacro:property name="rear_steering_damping" value="3.0" />
<xacro:property name="rear_steering_mu" value="0.001" />
<xacro:property name="rear_steering_friction" value="${rear_steering_mu * gravitational_acceleration * 0.25*vehicle_body_mass}" />

<!-- wheelの接地面の摩擦などを定義 -->
<xacro:property name="wheel_surface_mu1" value="0.8"/>
<xacro:property name="wheel_surface_mu2" value="0.8" />
<xacro:property name="wheel_kp" value="5e4" /> 
<xacro:property name="wheel_kd" value="1.0" />
<xacro:property name="wheel_minDepth" value="0.01" />
<xacro:property name="wheel_maxVel" value="1.0" />

<!-- rotationは共通 -->
<xacro:property name="rotational_joint_damping" value="0.01" />
<xacro:property name="rotational_joint_mu" value="0.001" />
<xacro:property name="rotational_joint_friction" value="${rotational_joint_mu * gravitational_acceleration}" />

<xacro:include filename="$(find 4ws_backstepping)/urdf/materials.xacro" />
<xacro:include filename="$(find 4ws_backstepping)/urdf/vehicle_4ws.trans" />
<xacro:include filename="$(find 4ws_backstepping)/urdf/vehicle_4ws.gazebo" />

<link name="base_link"/>

<link name="body_link">
  <inertial>
    <origin xyz="0.001411408730126027 0.0 -0.05" rpy="0 0 0"/>
    <mass value="${vehicle_body_mass}"/>
    <inertia ixx="${vehicle_inertia_coef*114.688091}" iyy="${vehicle_inertia_coef*336.418835}" izz="${vehicle_inertia_coef*418.647558}" ixy="${vehicle_inertia_coef*0.010139}" iyz="${vehicle_inertia_coef*0.002956}" ixz="${-vehicle_inertia_coef*0.33997}"/>
  </inertial>
  <visual>
    <origin xyz="0 0 -0.376" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/base_link.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="0 0.0 -0.376" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/base_link.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="v1_rear_1">
  <inertial>
    <origin xyz="2.514264385022891e-07 -0.05 -0.014912246259012996" rpy="0 0 0"/>
    <mass value="12.646385127140846"/>
    <inertia ixx="0.752015" iyy="0.005192" izz="0.755318" ixy="0.0" iyz="0.0" ixz="-0.0"/>
  </inertial>
  <visual>
    <origin xyz="0.44955 -0.05752 -0.376" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/v1_rear_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="0.44955 -0.05752 -0.376" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/v1_rear_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="v1_front_1">
  <inertial>
    <origin xyz="2.5142643794717756e-07 -0.05 -0.014912246259013162" rpy="0 0 0"/>
    <mass value="12.64638512714101"/>
    <inertia ixx="0.752015" iyy="0.005192" izz="0.755318" ixy="-0.0" iyz="0.0" ixz="-0.0"/>
  </inertial>
  <visual>
    <origin xyz="-0.45045 -0.05752 -0.376" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/v1_front_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="-0.45045 -0.05752 -0.376" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/v1_front_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="new_hinge_1">
  <inertial>
    <origin xyz="-6.714468377522209e-05 -0.05 -0.11523004751915897" rpy="0 0 0"/>
    <mass value="4.884144703315957"/>
    <inertia ixx="0.039519" iyy="0.032652" izz="0.021551" ixy="0.0" iyz="0.0" ixz="-3e-05"/>
  </inertial>
  <visual>
    <origin xyz="-0.45045 0.34348 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="-0.45045 0.34348 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="new_hinge_3">
  <inertial>
    <origin xyz="-6.714468377522209e-05 -0.05 -0.11523004751915977" rpy="0 0 0"/>
    <mass value="4.884144703315957"/>
    <inertia ixx="0.039519" iyy="0.032652" izz="0.021551" ixy="-0.0" iyz="0.0" ixz="-3e-05"/>
  </inertial>
  <visual>
    <origin xyz="-0.45045 -0.45752 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_3.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="-0.45045 -0.45752 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_3.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="new_hinge_2">
  <inertial>
    <origin xyz="-6.714468377322369e-05 -0.05 -0.1152300475191588" rpy="0 0 0"/>
    <mass value="4.884144703315957"/>
    <inertia ixx="0.039519" iyy="0.032652" izz="0.021551" ixy="0.0" iyz="0.0" ixz="-3e-05"/>
  </inertial>
  <visual>
    <origin xyz="0.44955 0.34348 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_2.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="0.44955 0.34348 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_2.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="new_hinge_4">
  <inertial>
    <origin xyz="-6.714468377405636e-05 -0.05 -0.11523004751915963" rpy="0 0 0"/>
    <mass value="4.884144703315957"/>
    <inertia ixx="0.039519" iyy="0.032652" izz="0.021551" ixy="0.0" iyz="0.0" ixz="-3e-05"/>
  </inertial>
  <visual>
    <origin xyz="0.44955 -0.45752 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_4.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="0.44955 -0.45752 -0.356" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_hinge_4.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<!-- ホイールのリンク設定 -->
<link name="new_wheel_1">
  <inertial>
    <origin xyz="3.1709373395083507e-05 -0.05 0.00021681028363659816" rpy="0 0 0"/>
    <mass value="4.639715839283939"/>
    <inertia ixx="0.029034" iyy="0.053334" izz="0.029034" ixy="-0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="-0.450031 0.34348 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <origin xyz="0 -0.05 0" rpy="1.57 0 0"/>
    <geometry>
      <cylinder radius="0.153" length="0.1"/>
    </geometry>

    <!-- ここから SDF の surface/friction -->
    <surface>
      <friction>
        <ode>
          <mu>${wheel_surface_mu1}</mu>
          <mu2>${wheel_surface_mu2}</mu2>
        </ode>
      </friction>
      <contact>
        <ode>
          <kp>${wheel_kp}</kp>
          <kd>${wheel_kd}</kd>
          <minDepth>${wheel_minDepth}</minDepth>
          <maxVel>${wheel_maxVel}</maxVel>
        </ode>
      </contact>
    </surface>
  </collision>
</link>

<link name="new_wheel_2">
  <inertial>
    <origin xyz="3.1709373397914575e-05 -0.05 0.00021681028363673693" rpy="0 0 0"/>
    <mass value="4.639715839283939"/>
    <inertia ixx="0.029034" iyy="0.053334" izz="0.029034" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0.449969 0.34348 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_2.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <!-- <origin xyz="0.449969 0.34348 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_2.stl" scale="0.001 0.001 0.001"/>
    </geometry> -->
    <origin xyz="0 -0.05 0" rpy="1.57 0 0"/>
    <geometry>
      <cylinder radius="0.153" length="0.1"/>
    </geometry>

    <!-- ここから SDF の surface/friction -->
    <surface>
      <friction>
        <ode>
          <mu>${wheel_surface_mu1}</mu>
          <mu2>${wheel_surface_mu2}</mu2>
        </ode>
      </friction>
      <contact>
        <ode>
          <kp>${wheel_kp}</kp>
          <kd>${wheel_kd}</kd>
          <minDepth>${wheel_minDepth}</minDepth>
          <maxVel>${wheel_maxVel}</maxVel>
        </ode>
      </contact>
    </surface>
  </collision>
</link>



<link name="new_wheel_3">
  <inertial>
    <origin xyz="3.1709373395083507e-05 -0.05 0.00021681028363579324" rpy="0 0 0"/>
    <mass value="4.639715839283939"/>
    <inertia ixx="0.029034" iyy="0.053334" izz="0.029034" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="-0.450031 -0.45752 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_3.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <!-- <origin xyz="-0.450031 -0.45752 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_3.stl" scale="0.001 0.001 0.001"/>
    </geometry> -->
    <origin xyz="0 -0.05 0" rpy="1.57 0 0"/>
    <geometry>
      <cylinder radius="0.153" length="0.1"/>
    </geometry>

    <!-- ここから SDF の surface/friction -->
    <surface>
      <friction>
        <ode>
          <mu>${wheel_surface_mu1}</mu>
          <mu2>${wheel_surface_mu2}</mu2>
        </ode>
      </friction>
      <contact>
        <ode>
          <kp>${wheel_kp}</kp>
          <kd>${wheel_kd}</kd>
          <minDepth>${wheel_minDepth}</minDepth>
          <maxVel>${wheel_maxVel}</maxVel>
        </ode>
      </contact>
    </surface>
  </collision>
</link>

<link name="new_wheel_4">
  <inertial>
    <origin xyz="3.170937339708191e-05 -0.05 0.00021681028363593202" rpy="0 0 0"/>
    <mass value="4.639715839283939"/>
    <inertia ixx="0.029034" iyy="0.053334" izz="0.029034" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0.449969 -0.45752 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_4.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="_"/>
  </visual>
  <collision>
    <!-- <origin xyz="0.449969 -0.45752 -0.150215" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://4ws_backstepping/new_meshes/new_wheel_4.stl" scale="0.001 0.001 0.001"/>
    </geometry> -->
    <origin xyz="0 -0.05 0" rpy="1.57 0 0"/>
    <geometry>
      <cylinder radius="0.153" length="0.1"/>
    </geometry>

    <!-- ここから SDF の surface/friction -->
    <surface>
      <friction>
        <ode>
          <mu>${wheel_surface_mu1}</mu>
          <mu2>${wheel_surface_mu2}</mu2>
        </ode>
      </friction>
      <contact>
        <ode>
          <kp>${wheel_kp}</kp>
          <kd>${wheel_kd}</kd>
          <minDepth>${wheel_minDepth}</minDepth>
          <maxVel>${wheel_maxVel}</maxVel>
        </ode>
      </contact>
    </surface>
  </collision>
</link>



<joint name="body" type="fixed">
  <origin xyz="0 -0.05 0.376" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="body_link"/>
</joint>

<joint name="v1_body_to_rear" type="fixed">
  <origin xyz="-0.44955 0.05752 0" rpy="0 0 0"/>
  <parent link="body_link"/>
  <child link="v1_rear_1"/>
</joint>

<joint name="v1_body_to_front" type="fixed">
  <origin xyz="0.45045 0.05752 0" rpy="0 0 0"/>
  <parent link="body_link"/>
  <child link="v1_front_1"/>
</joint>

<joint name="front_right_steering" type="revolute">
  <origin xyz="0.0 -0.4 -0.02" rpy="0 0 0"/>
  <parent link="v1_front_1"/>
  <child link="new_hinge_1"/>
  <dynamics damping="${front_steering_damping}" friction="${front_steering_friction}"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.57" lower="-1.57" effort="200" velocity="200"/>
</joint>

<joint name="front_right_wheel" type="continuous">
  <origin xyz="-0.000419 0.0 -0.205785" rpy="0 0 0"/>
  <parent link="new_hinge_1"/>
  <child link="new_wheel_1"/>
  <dynamics damping="${front_wheel_joint_damping}" friction="${front_wheel_joint_friction}"/>
  <limit effort="500" velocity="500"/>
  <axis xyz="0.0 1.0 -0.0"/>
</joint>

<joint name="rear_right_wheel" type="continuous">
  <origin xyz="-0.000419 0.0 -0.205785" rpy="0 0 0"/>
  <parent link="new_hinge_2"/>
  <child link="new_wheel_2"/>
  <dynamics damping="${rear_wheel_joint_damping}" friction="${rear_wheel_joint_friction}"/>
  <limit effort="500" velocity="500"/>
  <axis xyz="0.0 1.0 -0.0"/>
</joint>

<joint name="front_left_steering" type="revolute">
  <origin xyz="0.0 0.4 -0.02" rpy="0 0 0"/>
  <parent link="v1_front_1"/>
  <child link="new_hinge_3"/>
  <dynamics damping="${front_steering_damping}" friction="${front_steering_friction}"/>
  <axis xyz="0.0 0.0 1.0"/>
  <limit upper="1.57" lower="-1.57" effort="200" velocity="200"/>
</joint>

<joint name="front_left_wheel" type="continuous">
  <origin xyz="-0.000419 0.0 -0.205785" rpy="0 0 0"/>
  <parent link="new_hinge_3"/>
  <child link="new_wheel_3"/>
  <dynamics damping="${front_wheel_joint_damping}" friction="${front_wheel_joint_friction}"/>
  <limit effort="500" velocity="500"/>
  <axis xyz="0.0 1.0 -0.0"/>
</joint>

<joint name="rear_left_wheel" type="continuous">
  <origin xyz="-0.000419 0.0 -0.205785" rpy="0 0 0"/>
  <parent link="new_hinge_4"/>
  <child link="new_wheel_4"/>
  <dynamics damping="${rear_wheel_joint_damping}" friction="${rear_wheel_joint_friction}"/>
  <limit effort="500" velocity="500"/>
  <axis xyz="0.0 1.0 -0.0"/>
</joint>

<!-- 後輪右ステア -->
<joint name="rear_right_steering" type="revolute">
  <origin xyz="0.0 -0.4 -0.02" rpy="0 0 0"/>
  <parent link="v1_rear_1"/>
  <child  link="new_hinge_2"/>
  <!-- ステア駆動時のダンピング・摩擦 -->
  <dynamics damping="${rear_steering_damping}" friction="${rear_steering_friction}"/>
  <!-- 回転軸は Z 方向 -->
  <axis    xyz="0.0 0.0 1.0"/>
  <!-- 可動範囲／トルク／速度制限 -->
  <limit  lower="-1.57" upper="1.57" effort="200" velocity="200"/>
</joint>

<!-- 後輪左ステア -->
<joint name="rear_left_steering" type="revolute">
  <origin xyz="0.0 0.4 -0.02" rpy="0 0 0"/>
  <parent link="v1_rear_1"/>
  <child  link="new_hinge_4"/>
  <dynamics damping="${rear_steering_damping}" friction="${rear_steering_friction}"/>
  <axis    xyz="0.0 0.0 1.0"/>
  <limit  lower="-1.57" upper="1.57" effort="200" velocity="200"/>
</joint>

</robot>
